clusterloader2/pkg/measurement/common/service_creation_latency.go

```
@@ -0,0 +1,317 @@
+	s.client = config.ClusterFramework.GetClientSets().GetClient()
+	action, err := util.GetString(config.Params, "action")
+	if err != nil {
+		return nil, err
+	}
+
+	switch action {
+	case "start":
+		if s.selector.Parse(config.Params); err != nil {
+			return nil, err
+		}
+		s.waitTimeout, err = util.GetDurationOrDefault(config.Params, "waitTimeout", defaultServiceCreationLatencyTimeout)
+		if err != nil {
+			return nil, err
+		}
+		return nil, s.start()
+	case "waitForReady":
+		err = s.waitForReady()
+		if err != nil {
+			klog.Errorf("%s: %v", s, err)
+		}
+		return nil, err
+	case "gather":
+		return s.gather()
+	default:
+		return nil, fmt.Errorf("unknown action %v", action)
+	}
+}
+
+// Dispose cleans up after the measurement.
+func (s *serviceCreationLatencyMeasurement) Dispose() {
+	if s.isRunning {
+		s.isRunning = false
+		close(s.stopCh)
+	}
+	s.queue.Stop()
+	s.pingCheckers.Dispose()
+}
+
+// String returns a string representation of the metric.
+func (s *serviceCreationLatencyMeasurement) String() string {
+	return serviceCreationLatencyName + ": " + s.selector.String()
+}
+
+func (s *serviceCreationLatencyMeasurement) start() error {
+	if s.isRunning {
+		klog.Infof("%s: wait for controlled pods measurement already running", s)
+		return nil
+	}
+	klog.Infof("%s: starting wait for controlled pods measurement...", s)
+
+	s.isRunning = true
+	s.stopCh = make(chan struct{})
+
+	i := informer.NewInformer(
+		s.client,
+		"services",
+		s.selector,
+		func(oldObj, newObj interface{}) {
+			f := func() {
+				s.handleObject(oldObj, newObj)
+			}
+			s.queue.Add(&f)
+		},
+	)
+	return informer.StartAndSync(i, s.stopCh, informerSyncTimeout)
+}
+
+func (s *serviceCreationLatencyMeasurement) waitForReady() error {
+	for {
+		created, ipAssigned, reachable := 0, 0, 0
+		select {
+		case <-time.After(s.waitTimeout):
+			return fmt.Errorf("waiting for services timeout")
```

Please add context to that it's for "s" measurement.

---

Shouldn't this be done by the caller?

---

No - you should return a meaningful error for yourself.